django/thunderstore/social/api/experimental/views/mod_manager.py

```diff
@@ -0,0 +1,112 @@
+from typing import Sequence, Type
+
+import requests
+import ulid2  # type: ignore
+from django.conf import settings
+from django.contrib.auth import get_user_model, login
+from django.http import HttpResponse
+from drf_yasg.utils import swagger_auto_schema  # type: ignore
+from pydantic import BaseModel
+from rest_framework import serializers
+from rest_framework.exceptions import AuthenticationFailed
+from rest_framework.permissions import BasePermission
+from rest_framework.request import Request
+from rest_framework.response import Response
+from rest_framework.views import APIView
+
+from thunderstore.social.api.experimental.views.complete_login import (
+    get_or_create_auth_user,
+)
+from thunderstore.social.permissions import OauthSharedSecretPermission
+from thunderstore.social.providers import UserInfoSchema
+
+User = get_user_model()
+
+
+class MmLoginRequestBody(serializers.Serializer):
+    jwt = serializers.CharField(label="Authorization token")
+
+
+class MmLoginResponseBody(serializers.Serializer):
+    session_id = serializers.CharField()
+    username = serializers.CharField()
+
+
+class OverwolfProfileSchema(BaseModel):
+    """
+    https://overwolf.github.io/api/profile#get-user-profile-via-token
+    """
+
+    avatar: str
+    nickname: str
+    username: str
+
+
+class ModManagerLoginApiView(APIView):
+    """
+    Login user with information from Overwolf API using the received JWT
+    """
+
+    permission_classes: Sequence[Type[BasePermission]] = [OauthSharedSecretPermission]
+
+    @swagger_auto_schema(
+        operation_id="experimental.auth.mod_manager",
+        request_body=MmLoginRequestBody,
+        responses={200: MmLoginResponseBody()},
+    )
+    def post(self, request: Request) -> HttpResponse:
+        request_data = MmLoginRequestBody(data=request.data)
+        request_data.is_valid(raise_exception=True)
+        jwt: str = request_data.validated_data["jwt"]
+
+        verify_overwolf_jwt(jwt)
+        ow_profile = get_overwolf_user_profile(jwt)
+        user_info = get_user_info(ow_profile)
+        user = get_or_create_auth_user(user_info)
+        login(request, user, "django.contrib.auth.backends.ModelBackend")
+
+        if not request.session.session_key:
```

Comment:
Can this actually happen? How does login even work without a session?

Reply:
From SO: _"If `request.session.session_key` is `None`, that means the session object is brand-new and hasn't been saved to the db yet."_ (And yes, I added it since the key was None, although I'm not 100% sure if that's just hiding some other problem?)

Reply:
I'd assume `login` writes data to the session, so I'm left wondering how can that work if a session doesn't exist yet. But perhaps it creates the object without saving it, and `.create()` will actually commit it to the DB?

So I suppose the conclusion is, if the generated session token works, this isn't a problem.

Reply:
FYI spent some more time checking this out. Seems the case where the session isn't saved at that point would be hitting [this condition](https://github.com/django/django/blob/main/django/contrib/auth/__init__.py#L106-L116). I'm like 99% certain I did hit it while developing (since I wouldn't added that if-check otherwise), but I can't seem to reproduce it anymore.

Anyway I think it's safer to leave that check there, just in case.